pkg/epp/datastore/datastore.go

```diff
@@ -57,56 +61,94 @@ type Datastore interface {
 	ModelGetAll() []*v1alpha2.InferenceModel
 
 	// PodMetrics operations
-	PodUpdateOrAddIfNotExist(pod *corev1.Pod) bool
-	PodUpdateMetricsIfExist(namespacedName types.NamespacedName, m *Metrics) bool
-	PodGet(namespacedName types.NamespacedName) *PodMetrics
+	// PodGetAll returns all pods and metrics, including fresh and stale.
+	PodGetAll() []backendmetrics.PodMetrics
+	// PodList lists pods matching the given predicate.
+	PodList(func(backendmetrics.PodMetrics) bool) []backendmetrics.PodMetrics
+	PodUpdateOrAddIfNotExist(pod *corev1.Pod, pool *v1alpha2.InferencePool) bool
```

Comment:
Passing in the pool as a parameter here and in PodResyncAll is odd, the store is defined on a specific pool, and so it must enforce that the pods and models in the pool are the ones that belong to the pool that the datastore is defined on.

Reply:
Agree it seems a bit odd to the first glance, but this forces the caller to make sure the pool is valid. Otherwise in PodUpdateOrAddIfNotExist, and PodResyncAll, we need to check whether the pool exists. Currently the code just assumes the pool exists and ignores the error, which I think is a dangerous assumption (we should never ignore error). 

But also open to not passing the pool and just add validation logic in the PodXX functions.

Reply:
+1 to adding those error checks and have those functions return an error, I agree this is better; passing the pool makes it look like any other pool can be used to refresh the pods, which the interface better to be designed in a way to not allow.

Reply:
Upon a second thought, I actually think passing the pool in makes a lot of sense, it reads like "add these pods for this pool". So this pushes the responsibility to the caller to make sure the pods and pool match. Otherwise the datastore needs to validate that (basically duplicating some logic in the pool reconciler). The datastore is already pretty busy, I would prefer the datastore to be as thin as possible and not deal with too much business logic.